lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -154,6 +155,97 @@ void PatchReadFirstLane::getAnalysisUsage(AnalysisUsage &analysisUsage) const {
   analysisUsage.setPreservesCFG();
 }
 
+// =====================================================================================================================
+// Use the uniform (non-divergent) node instead of a divergent one if they are equal.
+//
+// Detect a case where a branch condition tests two nodes for equality where one of them
+// is divergent and the other uniform. In the true block replace the divergent
+// use with the uniform one as it helps instruction selection decide what to put in an SGPR.
+//
+// For example, in the code snippet below in BB1, %divergent can be replaced with %uniform.
+//
+// BB0:
+// %uniform = call i32 @llvm.amdgcn.readlane(i32 %divergent, i32 %lane)
+// %cmp = icmp eq i32 %divergent, %uniform
+// br i1 %cmp, label %BB1, label %BB2
+//
+// BB1:
+// ..
+// %use = zext i32 %divergent to i64 ; can use %uniform
+//
+// @param [in,out] function : LLVM function to be run for the optimization.
+bool PatchReadFirstLane::promoteEqualUniformOps(Function &function) {
+
+  bool changed = false;
+  for (auto &bb : function) {
+    // Look for a block that has a single predecessor with conditional branch.
+    const auto *pred = bb.getSinglePredecessor();
+    if (!pred)
+      continue;
+    const auto *branchInst = dyn_cast<BranchInst>(pred->getTerminator());
+    if (!branchInst || !branchInst->isConditional())
+      continue;
+    assert(branchInst->getSuccessor(0) != branchInst->getSuccessor(1));
+
+    // Make sure the condition for entering bb is the equality test passing.
+    bool match = false;
+    const auto *cmpInst = dyn_cast<Instruction>(branchInst->getCondition());
+    if (const auto *iCmpCond = dyn_cast_or_null<ICmpInst>(cmpInst)) {
+      switch (iCmpCond->getPredicate()) {
+      case ICmpInst::ICMP_EQ:
+        match = branchInst->getSuccessor(0) == &bb;
+        break;
+      case ICmpInst::ICMP_NE:
+        match = branchInst->getSuccessor(1) == &bb;
+        break;
+      default:
+        break;
+      }
+    } else if (const auto *fCmpCond = dyn_cast_or_null<FCmpInst>(cmpInst)) {
+      switch (fCmpCond->getPredicate()) {
+      case FCmpInst::FCMP_OEQ:
```

Comment:
Do we ever care about signed zeros? If so, it is not safe to replace uses of one value with another just because they are floating-point-equal, because they could be zeros with different signs.

Reply:
Good point. Do you agree that the right fix would be to do the conversion only if "nsz" FMF flag is present on fcmp?

Reply:
Yes I think testing for the nsz flag on the fcmp is probably fine. (Technically it might be more correct to check for the nsz flag on every use that you are going to replace, but that's much more work, and in practice if we add the nsz flag in one place then we add it everywhere.)

It would also be safe to do the optimization if you can prove that either operand of the comparison is non-zero, e.g. if one operand is a non-zero constant.

Reply:
I took a closer look at the patterns coming from real world applications and fcmp pattern would have _zero_ hits. Therefore I am removing it altogether - there is no justification to keep it.